tests/test_transpiler_unitary_decompositions.py

```diff
@@ -0,0 +1,206 @@
+import numpy as np
+import pytest
+from scipy.linalg import expm
+
+from qibo import gates, matrices
+from qibo.backends import NumpyBackend
+from qibo.models import Circuit
+from qibo.transpiler.unitary_decompositions import (
+    bell_basis,
+    calculate_h_vector,
+    calculate_psi,
+    calculate_single_qubit_unitaries,
+    cnot_decomposition,
+    cnot_decomposition_light,
+    magic_basis,
+    magic_decomposition,
+    to_bell_diagonal,
+    two_qubit_decomposition,
+)
+
+NREPS = 3  # number of repetitions to execute random tests
+ATOL = 1e-12
+
+
+def random_state(nqubits):
+    shape = 2**nqubits
+    psi = np.random.random(shape) + 1j * np.random.random(shape)
+    return psi / np.sqrt(np.sum(np.abs(psi) ** 2))
+
+
+def random_unitary(nqubits):
+    """Generates a random unitary matrix acting on nqubits."""
+    shape = 2 * (2**nqubits,)
+    m = np.random.random(shape) + 1j * np.random.random(shape)
+    return expm(1j * (m + np.conj(m.T)))
+
+
+def bell_unitary(hx, hy, hz):
+    ham = (
+        hx * np.kron(matrices.X, matrices.X)
+        + hy * np.kron(matrices.Y, matrices.Y)
+        + hz * np.kron(matrices.Z, matrices.Z)
+    )
+    return expm(-1j * ham)
+
+
+def purity(state):
```

Comment:
`purity` is also built-in in `quantum_info.metrics`

Reply:
I have tried using the built in function but in that case i get an error.

Reply:
Resolved